### PR TITLE
Integrate notifications to trigger privilege cache renewal into relevant workflows

### LINKED ISSUE #612: Triggers for marking RPT tokens as stale
Reference: https://openshift.io/openshiftio/Openshift_io/plan/detail/440

Basically: identify all source of triggers and call the service method that will be provided in #544.

### PR DESCRIPTION
**Pull request by sbryzak**
Fixes #612 

This PR completes the authorization/RPT feature by triggering re-evaluation of the privilege cache at key points of the workflow in various authorization-related operations.

### CODE REVIEW

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-5`

File: authorization/role/repository/identity_role.go [hunk 1]
```diff
@@ -525,7 +527,8 @@ func (m *GormIdentityRoleRepository) FindScopesByIdentityAndResource(ctx context
 	    FROM
 	      resource
 	    WHERE
-	      resource_id = ? /* RESOURCE_ID */
+          deleted_at IS NULL
```
Comment (sbose78) [inline authorization/role/repository/identity_role.go]: Good catch! gorm does it automatically, whereas in raw queries we gotta do it explicitly.

File: authorization/role/service/role_management_service.go [hunk 2]
```diff
@@ -139,8 +139,14 @@ func (s *roleManagementServiceImpl) Assign(ctx context.Context, assignedBy uuid.
...
-			for _, roleID := range existingRoleIDs {
-				err = s.Repositories().IdentityRoleRepository().Delete(ctx, roleID)
+			for _, ir := range existingRoles {
+				err = s.Repositories().IdentityRoleRepository().Delete(ctx, ir.IdentityRoleID)
+				if err != nil {
+					return err
+				}
+
+				// Notify the privilege cache service that an identity role has been deleted
+				err = s.Services().PrivilegeCacheService().NotifyIdentityResourcePrivilegesUpdated(ctx, ir.IdentityID, ir.ResourceID)
```
Comment (sbose78) [inline authorization/role/service/role_management_service.go]: Nice and clean way to update the cache!

File: authorization/role/service/role_management_service_blackbox_test.go [hunk 3]
```diff
@@ -483,6 +483,112 @@ func (s *roleManagementServiceBlackboxTest) TestRevokeResourceRolesOK() {
...
+	require.NoError(s.T(), err)
+	// At this stage the admin user should have no scopes
+	require.Len(s.T(), privs.ScopesAsArray(), 0)
+
+	// Assign the admin user the admin role for the resource, using the ForceAssign() function
+	err = s.Application.RoleManagementService().ForceAssign(s.Ctx, admin.IdentityID(), "admin", *res.Resource())
+	require.NoError(s.T(), err)
+
+	// Now the admin user should have the manage scope
+	privs, err = s.Application.PrivilegeCacheService().CachedPrivileges(s.Ctx, admin.IdentityID(), res.ResourceID())
```
Comment (sbose78) [inline authorization/role/service/role_management_service_blackbox_test.go]: We would need to test verify if the cache has been marked stale.

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-6`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-7`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-8`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-10`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-11`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-12`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-13`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-14`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-637-15`

File: authorization/invitation/service/invitation_service.go [hunk 4]
```diff
@@ -381,76 +382,95 @@ func (s *invitationServiceImpl) Accept(ctx context.Context, token uuid.UUID) (st
...
-		if err != nil {
-			return "", redirectOnFailure, errs.Wrapf(err, "failed to load identity for inviteTo %d", inviteTo)
-		}
+	var resourceID string
 
-		// If the invitation is for a membership, add a membership record
-		if inv.Member {
-			s.Repositories().Identities().AddMember(ctx, inviteToIdentity.ID, currentIdentityID)
-		}
+	err = s.ExecuteInTransaction(func() error {
```
Comment (xcoulon) [inline authorization/invitation/service/invitation_service.go]: the change here is because of a lack of transaction demarcation, isn't it ? Shouldn't it be part of another PR?

File: authorization/invitation/service/invitation_service.go [hunk 5]
```diff
@@ -381,76 +382,95 @@ func (s *invitationServiceImpl) Accept(ctx context.Context, token uuid.UUID) (st
...
+				if err != nil {
+					return err
+				}
 			}
 
-			err = s.Repositories().IdentityRoleRepository().Create(ctx, ir)
+			roles, err := s.Repositories().InvitationRepository().ListRoles(ctx, inv.InvitationID)
 			if err != nil {
-				return "", redirectOnFailure, errs.Wrap(err, "failed to create identity role")
+				return err
```
Comment (xcoulon) [inline authorization/invitation/service/invitation_service.go]: also, the error could be wrapped with a message to given some context information of where/when the `ListRoles` method was called

File: authorization/invitation/service/invitation_service.go [hunk 6]
```diff
@@ -381,76 +382,95 @@ func (s *invitationServiceImpl) Accept(ctx context.Context, token uuid.UUID) (st
...
+					IdentityID: currentIdentityID,
+					RoleID:     role.RoleID,
+					ResourceID: inviteToIdentityResourceID,
+				}
 
-		// Return the identity ID and redirect Path
-		return inviteToIdentityResourceID, redirectOnSuccess, nil
+				err = s.Repositories().IdentityRoleRepository().Create(ctx, ir)
+				if err != nil {
+					return err
```
Comment (xcoulon) [inline authorization/invitation/service/invitation_service.go]: same here for wrapping the error

File: authorization/permission/repository/privilege_cache.go [hunk 7]
```diff
@@ -139,16 +139,7 @@ func (m *GormPrivilegeCacheRepository) Create(ctx context.Context, privilegeCach
...
-	if err != nil {
-		log.Error(ctx, map[string]interface{}{
-			"privilege_cache_id": privilegeCache.PrivilegeCacheID,
-			"err":                err,
-		}, "unable to update privilege cache")
-		return errs.WithStack(err)
-	}
-
-	err = m.db.Model(obj).Updates(privilegeCache).Error
+	err := m.db.Save(privilegeCache).Error
```
Comment (xcoulon) [inline authorization/permission/repository/privilege_cache.go]: good point, there's no need to check if the record already exists ;)

File: authorization/permission/service/privilege_cache_service.go [hunk 8]
```diff
@@ -67,7 +67,10 @@ func (s *privilegeCacheServiceImpl) CachedPrivileges(ctx context.Context, identi
 				ExpiryTime: time.Now().Local().Add(time.Second * time.Duration(s.conf.GetPrivilegeCacheExpirySeconds())),
 			}
 
-			err = s.Repositories().PrivilegeCacheRepository().Create(ctx, privilegeCache)
+			err = s.ExecuteInTransaction(func() error {
```
Comment (xcoulon) [inline authorization/permission/service/privilege_cache_service.go]: what about having the whole method body in a single transaction ?

Comment (xcoulon) [reply authorization/permission/service/privilege_cache_service.go]: also, what about making the logic simpler, using the GORM `save` method to either insert or update the record, after the change you made in https://github.com/fabric8-services/fabric8-auth/pull/637/files#diff-55ea2c2abd467421bda2d07c8642a7f7R139?

File: authorization/role/repository/identity_role.go [hunk 9]
```diff
@@ -654,3 +697,136 @@ func identityRoleFilterByIdentityID(identityID uuid.UUID) func(db *gorm.DB) *gor
...
 	}
 }
+
+// FlagStaleForIdentityRoleChange executes two update queries; the first sets the stale flag to true for all privilege cache records where
+// the identity ID is equal to, or a descendent of (via memberships) the specified identity ID, and the resourceID is
+// equal to, or a descendent of (via the resource hierarchy) the specified resource ID.
+// The second query updates the token table, setting the STALE flag of the token STATUS field to true, for all
+// token records that are mapped to the corresponding privilege cache records in the first query, via the
+// many-to-many TOKEN_PRIVILEGE table
+func (m *GormIdentityRoleRepository) FlagPrivilegeCacheStaleForIdentityRoleChange(ctx context.Context, identityID uuid.UUID, resourceID string) error {
```
Comment (xcoulon) [inline authorization/role/repository/identity_role.go]: same question as above about the location of this method here, instead of in the privilege cache repository.

File: authorization/role/service/role_management_service.go [hunk 10]
```diff
@@ -192,7 +192,12 @@ func (s *roleManagementServiceImpl) ForceAssign(ctx context.Context, assignedTo
 			RoleID:     role.RoleID,
 		}
 
-		return s.Repositories().IdentityRoleRepository().Create(ctx, &ir)
+		err = s.Repositories().IdentityRoleRepository().Create(ctx, &ir)
+		if err != nil {
+			return err
```
Comment (xcoulon) [inline authorization/role/service/role_management_service.go]: could you wrap the `err` ?

File: authorization/role/service/role_management_service.go [hunk 11]
```diff
@@ -87,7 +87,7 @@ func (s *roleManagementServiceImpl) Assign(ctx context.Context, assignedBy uuid.
 	// privileges for the resource, otherwise the invitation workflow should be used instead
 	assignments := make(map[uuid.UUID][]uuid.UUID)
 
-	var existingRoleIDs []uuid.UUID
+	existingRoles := []rolerepo.IdentityRole{}
```
Comment (xcoulon) [inline authorization/role/service/role_management_service.go]: just curious, what's the benefit of an array of `IdentityRole` vs `UUID` ?

File: authorization/role/service/role_management_service_blackbox_test.go [hunk 12]
```diff
@@ -483,6 +483,112 @@ func (s *roleManagementServiceBlackboxTest) TestRevokeResourceRolesOK() {
 	assert.Len(s.T(), idRoles, 10)
 }
 
+func (s *roleManagementServiceBlackboxTest) TestPrivilegeCacheNotified() {
```
Comment (xcoulon) [inline authorization/role/service/role_management_service_blackbox_test.go]: should you also verify that another member of the same team does not have her cache marked as stale when someone else joins/leaves the team ?

File: authorization/token/service/token_service.go [hunk 13]
```diff
@@ -233,15 +233,27 @@ func (s *tokenServiceImpl) Audit(ctx context.Context, identity *account.Identity
...
-				perm := &token.Permissions{
-					ResourceSetID: &oldPrivResourceID,
-					Scopes:        oldPriv.ScopesAsArray(),
-					Expiry:        oldPriv.ExpiryTime.Unix(),
+				// If the old privilege is stale, then refresh its scopes and expiry time
+				if oldPriv.Stale {
+					// Lookup the scopes for the old privilege, as they may have changed
+					privilegeCache, err = s.Services().PrivilegeCacheService().CachedPrivileges(ctx, identity.ID, oldPrivResourceID)
+					if err != nil {
+						return errors.NewInternalError(ctx, err)
```
Comment (xcoulon) [inline authorization/token/service/token_service.go]: unless the underlying error should explicitly become an `Internal Error`, we should just `wrap` it with a message about the context of the call

File: authorization/token/service/token_service_blackbox_test.go [hunk 14]
```diff
@@ -491,3 +491,306 @@ func (s *tokenServiceBlackboxTest) TestAuditNonExistentResource() {
 	require.Error(s.T(), err)
 	require.IsType(s.T(), err, errors.BadParameterError{})
 }
+
+func (s *tokenServiceBlackboxTest) TestTokenUpdatedWhenUserAcceptsResourceInvitation() {
```
Comment (xcoulon) [inline authorization/token/service/token_service_blackbox_test.go]: could you split the test if you cover multiple cases?

File: account/repository/identity.go [hunk 15]
```diff
@@ -570,10 +573,260 @@ func (m *GormIdentityRepository) AddMember(ctx context.Context, identityID uuid.
...
+UPDATE privilege_cache SET
+  STALE = true
+WHERE
+  resource_id IN (SELECT resource_id FROM resource_hierarchy)
+  AND identity_id IN (SELECT identity_id FROM member_identity_hierarchy)
+  AND deleted_at IS NULL
+  `, memberID, memberID, memberOf, memberOf)
+
+	if result.Error != nil {
+		return errors.NewInternalError(ctx, result.Error)
```
[the remark this answers is not in the document]
Comment (sbryzak) [reply account/repository/identity.go]: With internal errors like this, my opinion is that there is generally little value in including an additional "something bad happened" message on top of whatever is returned by the underlying cause.  We already get the stack, which tells us exactly where the error occurred, plus a message from the cause which is far more useful.

File: authorization/invitation/service/invitation_service.go (hunk 4 above)
Comment (sbryzak) [reply authorization/invitation/service/invitation_service.go]: If it was a significant change I would agree, however I think that including a [rider](https://en.wikipedia.org/wiki/Rider_(legislation)) like this in unrelated PR is fine if it is something minor and doesn't benefit from the overhead of creating another issue, PR and review.  I would hate for our agility to suffer due to overbearing process, especially when something like this is already subject to peer review anyway.

File: authorization/invitation/service/invitation_service.go (hunk 5 above)
Comment (sbryzak) [reply authorization/invitation/service/invitation_service.go]: See previous response re. error handling.

File: authorization/invitation/service/invitation_service.go (hunk 6 above)
Comment (sbryzak) [reply authorization/invitation/service/invitation_service.go]: See previous.

File: authorization/permission/repository/privilege_cache.go (hunk 7 above)
Comment (sbryzak) [reply authorization/permission/repository/privilege_cache.go]: This change was only made because Gorm doesn't support updates for boolean values.  See [this issue](https://github.com/jinzhu/gorm/issues/202) as one example (there are others).

File: authorization/permission/service/privilege_cache_service.go (hunk 8 above)
Comment (sbryzak) [reply authorization/permission/service/privilege_cache_service.go]: The transaction is only required if there's a DML operation required.  Gorm's `Create()` operation is only used to create a new record, while `Save()` is used to update all fields of a record.

File: authorization/role/repository/identity_role.go (hunk 9 above)
Comment (sbryzak) [reply authorization/role/repository/identity_role.go]: See previous response.

File: authorization/role/service/role_management_service.go (hunk 11 above)
Comment (sbryzak) [reply authorization/role/service/role_management_service.go]: No benefit, simply that `IdentityRoleRepository.FindIdentityRolesByIdentityAndResource()` returns an array of `IdentityRole`.

File: authorization/role/service/role_management_service.go (hunk 10 above)
Comment (sbryzak) [reply authorization/role/service/role_management_service.go]: See previous response.

File: authorization/role/service/role_management_service_blackbox_test.go (hunk 12 above)
Comment (sbryzak) [reply authorization/role/service/role_management_service_blackbox_test.go]: That's a good idea, I'll update the test to do that.

File: authorization/token/service/token_service.go (hunk 13 above)
Comment (sbryzak) [reply authorization/token/service/token_service.go]: Good catch, the underlying error here will already be an `InternalError` (as that is all the `CachedPrivileges()` function can return) so no need to wrap it again.

File: authorization/token/service/token_service_blackbox_test.go (hunk 14 above)
Comment (sbryzak) [reply authorization/token/service/token_service_blackbox_test.go]: This is a complex test that continuously builds upon and depends on state initialized by the previous steps.  It's not really feasible to split it up.

File: account/repository/identity.go [hunk 16]
```diff
@@ -570,10 +573,260 @@ func (m *GormIdentityRepository) AddMember(ctx context.Context, identityID uuid.
...
 	}
+
+	err = m.FlagPrivilegeCacheStaleForMembershipChange(ctx, identityID, memberID)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"member_of": identityID,
+			"member_id": memberID,
+			"err":       err,
+		}, "unable to create the membership - error notifying privilege cache")
+		return errs.WithStack(err)
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply account/repository/identity.go]: ok, let's discuss about the error/logging practices somewhere else.

File: account/repository/identity.go [hunk 17]
```diff
@@ -570,10 +573,260 @@ func (m *GormIdentityRepository) AddMember(ctx context.Context, identityID uuid.
...
+	}
+
+	err := m.db.Delete(membership).Error
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"member_of": memberOf,
+			"member_id": memberID,
+			"err":       err,
+		}, "unable to remove the membership")
+		return errs.WithStack(err)
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply account/repository/identity.go]: pk ;)

File: account/repository/identity.go [hunk 18]
```diff
@@ -570,10 +573,260 @@ func (m *GormIdentityRepository) AddMember(ctx context.Context, identityID uuid.
...
+// FlagStaleForMembershipChange executes two update queries; the first sets the stale flag to true for all privilege
+// cache records where the identity ID is equal to, or a descendent of (via memberships) the specified member ID, and
+// the resourceID is contained in a set of resources for which there is an IDENTITY_ROLE record for the resource, or
+// any of its descendent resources, and the IDENTITY_ROLE's identity is in the identity ancestor hierarchy specified by
+// the memberOf parameter.
+//
+// The second query updates the token table, setting the STALE flag of the token STATUS field to true, for all
+// token records that are mapped to the corresponding privilege cache records in the first query, via the
+// many-to-many TOKEN_PRIVILEGE table
+func (m *GormIdentityRepository) FlagPrivilegeCacheStaleForMembershipChange(ctx context.Context, memberID uuid.UUID, memberOf uuid.UUID) error {
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply account/repository/identity.go]: ok, if by design repositories are not supposed to call each other's methods.

File: account/repository/identity.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply account/repository/identity.go]: as said above, let's discuss about errors outside of the scope of this PR

File: authorization/invitation/service/invitation_service.go (hunk 4 above)
Comment (xcoulon) [reply authorization/invitation/service/invitation_service.go]: ok, fair enough (and thanks for the link to explain the `rider`)

File: authorization/role/service/role_management_service.go (hunk 11 above)
Comment (xcoulon) [reply authorization/role/service/role_management_service.go]: ok

File: authorization/role/service/role_management_service.go (hunk 10 above)
Comment (xcoulon) [reply authorization/role/service/role_management_service.go]: yes ;)

File: authorization/role/service/role_management_service_blackbox_test.go (hunk 12 above)
Comment (xcoulon) [reply authorization/role/service/role_management_service_blackbox_test.go]: thanks!

File: authorization/token/service/token_service_blackbox_test.go (hunk 14 above)
Comment (xcoulon) [reply authorization/token/service/token_service_blackbox_test.go]: ok, I did not realize that the previous steps mattered.

**Comment by xcoulon** [comment]
@sbryzak I approved your PR but @sbose78 may still have a pending request in https://github.com/fabric8-services/fabric8-auth/pull/637/files#discussion_r215133242, unless it's been addressed in some tests already?
